cub/cub/detail/fast_modulo_division.cuh

```diff
@@ -0,0 +1,221 @@
+ * larger_unsigned_type
+ **********************************************************************************************************************/
+
+template <typename T, typename = void>
+struct larger_unsigned_type
+{
+  static_assert(sizeof(T) >= 8, "64-bit integer are only supported from CUDA >= 11.5");
+  using type = void;
+};
+
+template <typename T>
+struct larger_unsigned_type<T, typename ::cuda::std::enable_if<(sizeof(T) < 4)>::type>
+{
+  using type = unsigned;
+};
+
+template <typename T>
+struct larger_unsigned_type<T, typename ::cuda::std::enable_if<(sizeof(T) == 4)>::type>
+{
+  using type = ::cuda::std::uint64_t;
+};
+
+#if CUB_IS_INT128_ENABLED
+
+template <typename T>
+struct larger_unsigned_type<T, typename ::cuda::std::enable_if<(sizeof(T) == 8)>::type>
+{
+  using type = __uint128_t;
+};
+
+#endif // CUB_IS_INT128_ENABLED
+
+template <typename T>
+using larger_unsigned_type_t = typename larger_unsigned_type<T>::type;
+
+template <typename T>
+using implicit_prom_t = decltype(+T{});
+
+template <typename T>
+using unsigned_implicit_prom_t = typename ::cuda::std::make_unsigned<implicit_prom_t<T>>::type;
+
+/***********************************************************************************************************************
+ * Extract higher bits after multiplication
+ **********************************************************************************************************************/
+
+template <typename T, typename R>
+_CCCL_NODISCARD _CCCL_HOST_DEVICE _CCCL_FORCEINLINE unsigned_implicit_prom_t<T>
+multiply_extract_higher_bits(T value, R multiplier)
+{
+  static_assert(::cuda::std::is_integral<T>::value && !::cuda::std::is_same<T, bool>::value, "unsupported type");
+  static_assert(::cuda::std::is_integral<R>::value && !::cuda::std::is_same<R, bool>::value, "unsupported type");
+  _CCCL_ASSERT(value >= 0, "value must be non-negative");
+  _CCCL_ASSERT(multiplier >= 0, "multiplier must be non-negative positive");
+  using unsigned_t           = unsigned_implicit_prom_t<T>;
+  using larger_t             = larger_unsigned_type_t<T>;
+  constexpr unsigned BitSize = sizeof(T) * CHAR_BIT;
+  // clang-format off
+  NV_IF_TARGET(
+    NV_IS_HOST,
+      (return static_cast<unsigned_t>((static_cast<larger_t>(value) * multiplier) >> BitSize);),
+    //NV_IS_DEVICE
+      (return (sizeof(T) == 8)
+        ? static_cast<unsigned_t>(__umul64hi(value, multiplier))
+        : static_cast<unsigned_t>((static_cast<larger_t>(value) * multiplier) >> BitSize);));
+  // clang-format on
+}
+
+/***********************************************************************************************************************
+ * Fast Modulo/Division based on Precomputation
+ **********************************************************************************************************************/
+
+template <typename T>
+struct fast_div_mod
```

Comment:
Q: Do we know that an optimizing compiler cannot perform the optimization hand-rolled in the implementation below? This is a fair amount of code to maintain.

Important: please add unit tests for this new facility.

Reply:
the compiler is able to automatically perform this optimization if the value is known at compile-time. Here the idea to compute the heavy part on the host and perform the efficient version of modulo division on the device.
I will add a new unit test